src/main.cpp

```diff
@@ -179,6 +188,168 @@ BOOL CtrlHandler(DWORD fdwCtrlType)
 
 
 
+////////////////////////////////////////////////////////////////////////////////
+// universalMain - Main startup logic for both standard running and as a service
+
+void universalMain()
+{
+#ifdef _WIN32
+	if (!SetConsoleCtrlHandler((PHANDLER_ROUTINE)CtrlHandler, TRUE))
+	{
+		LOGERROR("Could not install the Windows CTRL handler!");
+	}
+#endif
+
+	// Initialize logging subsystem:
+	cLogger::InitiateMultithreading();
+
+	// Initialize LibEvent:
+	cNetworkSingleton::Get();
+
+#if !defined(ANDROID_NDK)
+	try
+#endif
+	{
+#if defined(_WIN32) || defined(_WIN64)
+		if (g_RunAsService)
+		{
+			cRoot::m_RunningAsService = true;
+		}
+#endif
+		Root.Start();
+	}
+#if !defined(ANDROID_NDK)
+	catch (std::exception & e)
+	{
+		LOGERROR("Standard exception: %s", e.what());
+	}
+	catch (...)
+	{
+		LOGERROR("Unknown exception!");
+	}
+#endif
+
+	g_ServerTerminated = true;
+
+	// Shutdown all of LibEvent:
+	cNetworkSingleton::Get().Terminate();
+}
+
+
+
+
+#if defined(_WIN32) || defined(_WIN64)
+////////////////////////////////////////////////////////////////////////////////
+// serviceWorkerThread: Keep the service alive
+
+DWORD WINAPI serviceWorkerThread(LPVOID lpParam)
+{
+	UNREFERENCED_PARAMETER(lpParam);
+
+	// Do the normal startup
+	universalMain();
+
+	return ERROR_SUCCESS;
+}
+
+
+
+
+////////////////////////////////////////////////////////////////////////////////
+// serviceSetState: Set the internal status of the service
+
+void serviceSetState(DWORD acceptedControls, DWORD newState, DWORD exitCode)
+{
+	SERVICE_STATUS serviceStatus;
+	ZeroMemory(&serviceStatus, sizeof(SERVICE_STATUS));
+	serviceStatus.dwCheckPoint = 0;
+	serviceStatus.dwControlsAccepted = acceptedControls;
+	serviceStatus.dwCurrentState = newState;
+	serviceStatus.dwServiceSpecificExitCode = 0;
+	serviceStatus.dwServiceType = SERVICE_WIN32;
+	serviceStatus.dwWaitHint = 0;
+	serviceStatus.dwWin32ExitCode = exitCode;
+
+	if (SetServiceStatus(g_StatusHandle, &serviceStatus) == FALSE)
+	{
+		LOGERROR("SetServiceStatus() failed\n");
+	}
+}
+
+
+
+
+////////////////////////////////////////////////////////////////////////////////
+// serviceCtrlHandler: Handle stop events from the Service Control Manager
+
+void WINAPI serviceCtrlHandler(DWORD CtrlCode)
+{
+	switch (CtrlCode)
+	{
+	case SERVICE_CONTROL_STOP:
+		Root.SetStopping(true);
+		serviceSetState(0, SERVICE_STOP_PENDING, 0);
+		break;
+
+	default:
+		break;
+	}
```

Comment:
We use braces for the insides of each `case`, and indent the `case` labels one level more.